### ISSUE TITLE
RTP-MIDI not working properly

### ISSUE THREAD

**Issue by wanthalf** [zynthian/zynthian-issue-tracking#831]
RTP-MIDI is not working out-of-the box. User @mbvs identified (and I can confirm) two issues in MacOS as described at https://discourse.zynthian.org/t/rtp-midi-problems/6619. But they do not really seem to be specific to MacOS:

- Connection with the ZYNTHIAN RTPMIDI only works if i disable IPv6 for the network connection on my mac. Otherwise I get an error that the zynthian didn’t respond to the connection attempt. Probably, zynthian's RTP-MIDI is just not listening on IPv6 at all?

- Only the `rtpmidi_out` -- `net_in` connection works by default (i.e. receiving messages from mac). To get sending working I need to connect `rtpmidi_in` to `net_out` manually in Patchage despite having jackrtpmidid:rtpmidi in checked under MIDI OUTPUT Ports in WebConfig.

**Comment by oscaracena** [comment]
I think that this issue may be related with what is discussed here:

* https://discourse.zynthian.org/t/control-driver-for-a-device-connected-with-rtp-midi-qmidinet/10758/10

**Comment by riban-bw** [comment]
@oscaracena you found the issue in rtpmidi and fixed it in a branch. Did this ever get pushed upstream and merged into zynthian?

**Comment by oscaracena** [comment]
> Did this ever get pushed upstream and merged into zynthian?

I don't think so, I you look a that [file/line](https://github.com/bbouchez/jackrtpmidid/blob/a646b3c6ffcc9cb9eaec47ae89bbcf9715eb46af/jackrtpmidid.cpp#L90) in the master branch, the code is the old version. I think @bbouchez was [very busy](https://discourse.zynthian.org/t/control-driver-for-a-device-connected-with-rtp-midi-qmidinet-network-midi-2-0/10758/17?u=oscaracena) at the time, and then it got diluted over time 😅

**Comment by bbouchez** [comment]
Oh gosh !!!!

I completely forgot about that problem !!!
I am so busy with Network UMP that RTP-MIDI has been pushed aside...
I will need to resynchronize myself before digging again the code, but clearly I have to correct it 😅

By the way, a small comment about the delta-time discussion in the Zynthian forum  : I know they are all set to 0, and I **strongly** recommend to keep it like that. Almost no application in the world is able to deal correctly with the delta-time if it is not 0, even the Mac OS one ! We have made plenty of tests with other RTP-MIDI devices with non-null delta times :
- at the best, the delta-times are simply ignored, but at least, you get expected events in the order they have set in the data block (but all sent in bursts)
- in the most case, the MIDI stream being reconstructed by the driver becomes totally corrupted with events not being played in the correct order or with erratic timing

**Comment by bbouchez** [comment]
After checking the original report from @wanthalf and comparing with @oscaracena work, I think these are different issues.

From @wanthalf report :
 
- yes, the Zynthian RTP-MIDI driver does not listen on IPv6, only IPv4. But I don't understand why the IPv6 should be deactivated on the host side (the Mac), as if the remote participant address is entered in IPv4 format, the Mac has no reason to try to reach it using IPv6. I will perform some tests on my Mac later (for the three coming weeks, I will be unable to work in my studio, as I am completely rebuilding the floor where the studio is located)

- for the automatic connection of jack to RTP-MIDI direction which is not working, I don't think the problem is located in jackrtpmidid. From what I understand, the JACK port is exposed properly and accept the connection if done manually. So I think the problem is more related to the configuration engine of the Zynthian. Can somebody check if it is still the case ?

From @oscaracena analysis :

- for the jack to RTP-MIDI broken communication : yes, it's my entire fault... 
As I only use the Zynthian as a receiver driven by external sequencer / keyboards, I thought that nobody would use the RTP-MIDI output (😅 <- me being ashamed to think that all Zynthianers would use the Zynthian like me... LOL)
I will check if I can merge directly the change from @oscaracena (I think it will work properly) in my code, but as I said I need to find first a way to do tests without my studio... otherwise it may take 2 or 3 weeks before I push the corrected code

**Comment by riban-bw** [comment]
I just enabled two zynthian with RTP-MIDI but I don't see a way to target one to the other. Presumably there needs to be some kind of address set so that the two know to pass data. Maybe there needs to be some work in zynthian to show available source / destination devices.

**Comment by bbouchez** [comment]
> I just enabled two zynthian with RTP-MIDI but I don't see a way to target one to the other. Presumably there needs to be some kind of address set so that the two know to pass data. Maybe there needs to be some work in zynthian to show available source / destination devices.

Hi Riban,

this can not work. To establish a RTP-MIDI communication between two machines, you need a session initiator and a session listener. The session initiator is in charge to send the invitation messages and initiate the clock synchronization process. Right now, jackrtpmidid does not implement the session initiator parameters (but the CRTP_MIDI class instantiated by jackrtpmidid supports it, you just need to pass the listener IP:port and set IsSessionIntiator parameter at true)
